[CIR][CIRGen] Add complex type and its CIRGen support

Lancern opened the pull request:
This PR adds `!cir.complex` type to model the `_Complex` type in C. It also contains support for its CIRGen.

In detail, this patch adds the following CIR types, ops, and attributes:

- The `!cir.complex` type is added to model the `_Complex` type in C. This type is parameterized with the type of the components of the complex number, which must be either an integer type or a floating-point type.
- The `#cir.complex` attribute is added to represent a literal value of `_Complex` type. It is a struct-like attribute that provides the real and imaginary part of the literal `_Complex` value.
- The `cir.complex.extract` op is added to extract the real and imaginary part of a value of `!cir.complex` type.

CIRGen support for the new complex type is also added. Note the implementation diverges from the original clang CodeGen, where expressions of complex types are handled differently from scalars and aggregates. Instead, this patch treats expressions of complex types as scalars, as such expressions can be simply lowered to a CIR value of `!cir.complex` type.

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,92 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
```
Can you please add an extra RUN line that builds this for C++?

Lancern replied:
Updated.

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,92 @@
+// RUN: FileCheck --input-file=%t.cir %s
+
+double _Complex float_complex_basic(void) {
+  double _Complex x = 2.0 + 3.0i;
```
I'm a bit confused with the approach here, seems a bit less ideal than what LLVM generates: https://godbolt.org/z/3Go78MMh1, looking at the CIR output:
```
    %2 = cir.const(#cir.fp<2.000000e+00> : !cir.double) : !cir.double loc(#loc5)
    %3 = cir.const(#cir.complex<#cir.fp<0.000000e+00> : !cir.double, #cir.fp<3.000000e+00> : !cir.double> : !cir.complex<!cir.double>) : !cir.complex<!cir.double> loc(#loc4)
    %4 = cir.cast(float_to_complex, %2 : !cir.double), !cir.complex<!cir.double> loc(#loc5)
    %5 = cir.binop(add, %4, %3) : !cir.complex<!cir.double> loc(#loc81)
```
Why doesn't `%3` already builds the complex? Seems off that CIRGen first creates `%2`, followed by bitcast + addition.

bcardosolopes replied:
Question still remains, sounds like we could do a better initialization job here. It's also fine if you plan to do this as incremental improvement, but I'm mostly looking to hear more about the plan here.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
````diff
@@ -986,6 +992,51 @@ def CmpOp : CIR_Op<"cmp", [Pure, SameTypeOperands]> {
+    Example:
+
+    ```mlir
+    !fcmp = !cir.complex<!cir.float>
````
`!fcmp` might sounds like "float comparison", perhaps use `!complex` to make the docs more readable.

Lancern replied:
Updated.

bcardosolopes reviewed (changes requested):
This is nice, thanks for working on it! 

Comments inline but overall I'm just a bit confused by the fact that LLVM skeleton isn't being much followed here (and some comments on CIRGen decisions).

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -986,6 +992,51 @@ def CmpOp : CIR_Op<"cmp", [Pure, SameTypeOperands]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+def ComplexExtractOp : CIR_Op<"complex.extract", [Pure]> {
```
Updated.

Lancern commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -357,6 +357,50 @@ LogicalResult cir::FPAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+  mlir::TypedAttr real;
+  mlir::TypedAttr imag;
+
+  if (elementTy.isa<mlir::cir::IntType>()) {
```
Updated.

bcardosolopes commented:
Thanks for the detailed explanation (very clarifying) and for thinking about code duplication!

> If we follow the skeleton, the code in CIRGenExprScalar.cpp and CIRGenExprComplex.cpp (which hasn't been invented) would be very very similar to each other, resulting in unnecessary redundancy. Thus I decided to treat expressions of complex types as scalar expressions in CIRGen. This eases the implementation greatly and it still follow the intuition.

`clang/lib/CodeGen/CGExprComplex.cpp` is quite big. Are you confident that current missing complex codegen will just apply with all scalar stuff? I have the feeling that as soon as something differs, we'll need to change scalar emitters to take complex types into account, and because we're assuming scalar handles it all, we won't have asserts to catch what's not currently implemented - so might end up with bad codegen which will take time to investigate and debug.

Unfortunately, I believe duplication will lead us to converge faster and have less issues. How about you add a `FIXME(cir): unlike traditional CodeGen, this shares common code with ScalarExprEmitter` to every method that is the same between `ComplexExprEmitter` and `ScalarExprEmitter`, so that when we complete complex support, we can decide if there's a refactoring approach we could take?

Lancern commented:
> How about you add a FIXME(cir): unlike traditional CodeGen, this shares common code with ScalarExprEmitter to every method that is the same between ComplexExprEmitter and ScalarExprEmitter, so that when we complete complex support, we can decide if there's a refactoring approach we could take?

OK I believe this is also acceptable and it allows us to evaluate this divergence more deeply before we can make further decisions. I'll move the complex CIRGen code to a new `CIRGenComplex.cpp` file and keep the upstream code structure for now.

Lancern commented:
Rebased onto the latest `main`, but not yet started to work on the reviews.

bcardosolopes commented:
@Lancern sorry, I was out on vacation. Let me know when you address all reviews and update the PR.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -171,6 +171,25 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
         CGF.getLoc(E->getExprLoc()), Ty,
         Builder.getAttr<mlir::cir::FPAttr>(Ty, E->getValue()));
   }
+  mlir::Value VisitImaginaryLiteral(const ImaginaryLiteral *E) {
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenCXX.cpp:
```diff
@@ -200,8 +200,6 @@ static void buildDeclInit(CIRGenFunction &CGF, const VarDecl *D,
     CGF.buildScalarInit(Init, CGF.getLoc(D->getLocation()), lv, false);
     return;
-  case TEK_Complex:
-    llvm_unreachable("complext evaluation NYI");
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenClass.cpp:
```diff
@@ -804,9 +804,6 @@ void CIRGenFunction::buildInitializerForField(FieldDecl *Field, LValue LHS,
     }
     break;
-  case TEK_Complex:
-    llvm_unreachable("NYI");
```
Updated.

Lancern commented:
@bcardosolopes Hi Bruno, I have updated the implementation so that it now follows the upstream skeleton.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprComplex.cpp:
```diff
@@ -0,0 +1,947 @@
+}
+
+mlir::Value ComplexExprEmitter::VisitInitListExpr(InitListExpr *E) {
+  if (E->getNumInits() == 2) {
```
Thanks for introducing the skeleton and for all the added testcases. However, this new file is quite big and introduces lots of CIRGen that I don't see tested (example, this one). Few options here:
- Add more testcases and exercise each of the visitors.
- Delete non-tested visitors and add `llvm_unreacheable` to them. Later on, do incremental patches to add more visitors and their correspondent tests.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenDecl.cpp:
```diff
@@ -684,6 +684,15 @@ void CIRGenFunction::buildScalarInit(const Expr *init, mlir::Location loc,
 
+void CIRGenFunction::buildComplexInit(const Expr *init, mlir::Location loc,
+                                      LValue lvalue) {
+  // TODO: this is where a lot of ObjC lifetime stuff would be done.
```
This needs an assert+unimplemented.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -242,6 +242,41 @@ def FPAttr : CIR_Attr<"FP", "fp", [TypedAttrInterface]> {
+    number value of the specified complex type.
+
+    The `real` parameter gives the real part of the complex number, and the
+    `imag` parameter gives the imaginary part of the complex number.
```
Add an example of it's use with `cir.const`.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -995,6 +1003,98 @@ def CmpOp : CIR_Op<"cmp", [Pure, SameTypeOperands]> {
+def ComplexCreateOp : CIR_Op<"complex.create", [Pure, SameTypeOperands]> {
+  let summary = "Create a new complex value from its real and imaginary parts";
+  let description = [{
+    The `cir.complex.create` operation takes two operands of the same type and
```
I don't see any use of this operation in the tests, please make sure it shows up.

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,243 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+// RUN: %clang_cc1 -x c++ -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
```
it would be nice to have a `test/CIR/IR/complex.cir` test for printing and parsing as well.

bcardosolopes reviewed (changes requested):
Very nice. This is in the right direction, thanks the updates. More comments inline!

bcardosolopes commented:
Let me know once this is ready for review again!

bcardosolopes commented:
@lanza rebased over the weekend, sorry for the churn. Can you rebase this PR again?

Lancern commented:
Rebased onto the latest `main`.

Lancern commented:
After some investigation, I believe I'd better start with something simpler. This big PR adds support for `_Complex` type directly, diverging from upstream clang CodeGen. Also it skips the `_Imaginary` type which is simpler and provides a reasonable foundation for the full `_Complex` type. So I'll start working on the `_Imaginary` type instead, and after that I'll come back to deal with the `_Complex` type.

Lancern commented:
@bcardosolopes Hi Bruno, I believe this PR is now ready for another round of review.

I decide to keep this PR simpler (but the size of the PR is still quite big). In the latest version I made the following adjustments:

- I replaced the `#cir.complex` attribute with the `#cir.imag` attribute. The former represents a whole complex number, while the latter represents a complex number that is purely imaginary (i.e. its real part is 0), which can be written in C as a number literal followed by the `i` suffix (e.g. `3.0i` will become `#cir.imag<#cir.fp<3.0>>`).
- I replaced the `cir.complex.real` and `cir.complex.imag` with `cir.complex.get_real` and `cir.complex.get_imag`, respectively. The main difference here is the `get_*` version yields pointers to the corresponding complex number component just like what `cir.get_member` is doing. This change is to support the `__real__` and `__imag__` operator which yield lvalues.
- I removed a lot of code in `CGExprComplex.cpp` and left most visitors as unimplemented for now. Only a few complex number operations are supported in this PR so code review can be easier, which include:
  - List initialization of complex numbers;
  - Imaginary literals;
  - Load and store of complex number values;
  - `__real__` and `__imag__` operator.

All supported operations are tested in `clang/test/CIR/CodeGen/complex.c` . Support for more complex number operations can be added in future PRs.

bcardosolopes reviewed (changes requested):
> I decide to keep this PR simpler (but the size of the PR is still quite big).

Thanks, I appreciate that.

> I replaced the #cir.complex attribute with the #cir.imag attribute. The former represents a whole complex number, while the latter represents a complex number that is purely imaginary (i.e. its real part is 0), which can be written in C as a number literal followed by the i suffix (e.g. 3.0i will become #cir.imag<#cir.fp<3.0>>).

After looking the PR, I'm not sure we need `#cir.imag`, it should just be an int or fp attribute directly.
```
//      CHECK:   %[[#REAL:]] = cir.const #cir.fp<1.000000e+00> : !cir.double
// CHECK-NEXT:   %[[#IMAG:]] = cir.const #cir.fp<2.000000e+00> : !cir.double
// CHECK-NEXT:   %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !cir.double -> !cir.complex<!cir.double>
```

This composes better with what `cir.complex.create` ends up doing (building imaginary out of regulaer fp/int directly).

> I replaced the cir.complex.real and cir.complex.imag with cir.complex.get_real and cir.complex.get_imag, respectively. The main difference here is the get_* version yields pointers to the corresponding complex number component just like what cir.get_member is doing. This change is to support the __real__ and __imag__ operator which yield lvalues.

I'll address this in another comment.

> I removed a lot of code in CGExprComplex.cpp and left most visitors as unimplemented for now. Only a few fundamental complex number operations are supported in this PR so code review can be easier, including

Awesome!

> All supported operations are tested in clang/test/CIR/CodeGen/complex.c . Support for more complex number operations can be added in future PRs.

Can you please add LLVM IR support and test it altogether in `complex.c`?

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir -o - %s | FileCheck %s --check-prefixes=C,CHECK
```
Please write files to disk, it makes it convenient to run locally and look at the output

Lancern replied:
Updated.

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+
+void imag_literal(VOID) {
+  c = 3.0i;
+  ci = 3i;
```
Here in `imag_literal()`
```
    %0 = cir.const #cir.imag<#cir.fp<3.000000e+00> : !cir.double> : !cir.complex<!cir.double>
    %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
    cir.store %0, %1 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```
How do you plan to distinguish if the `cir.store` is storing to real or imag part when lowering this? Relying on the imag attribute sounds brittle. Another reason to add LLVM lowering as part of the work, it'd have popped up this kind of question / design issue. This should be using the same mechanism as elsewhere:

```
    %0 = cir.const #cir.imag<#cir.fp<3.000000e+00> : !cir.double> : !cir.complex<!cir.double>
    %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
    %3 = cir.complex.get_img %1 : !cir.ptr<!cir.complex<!cir.double>> -> !cir.ptr<!cir.double>
    cir.store %0, %3 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```

Please rename `cir.complex.get_img` to `cir.complex.imag_ptr`, `cir.complex.get_real` to `cir.complex.real_ptr`.

Lancern replied:
As I have removed `#cir.imag` the CIRGen scheme under discussion here now becomes:

```
c = 3.0i;
// %0 = cir.const #cir.fp<3.000000e+00> : !cir.double
// %1 = cir.get_global @c : !cir.ptr<!cir.complex<!cir.double>>
// %2 = cir.complex.get_img %1 : !cir.ptr<!cir.complex<!cir.double>> -> !cir.ptr<!cir.double>
// cir.store %0, %2 : !cir.complex<!cir.double>, !cir.ptr<!cir.complex<!cir.double>>
```

However this may be a bit hard to implement during CIRGen since this assignment is basically an assignment between two complex value. The upstream CodeGen tracks the real and imaginary parts of a complex value individually so it can generate such code directly. In CIR we may use some canonicalization passes to do this job?

bcardosolopes replied:
Since a lot has changed since, let's ignore my first comment on this sub-thread. For the test on `c = 3.0i`, we now get:

```
//      CHECK: %[[#REAL:]] = cir.const #cir.fp<0.000000e+00> : !cir.double
// CHECK-NEXT: %[[#IMAG:]] = cir.const #cir.fp<3.000000e+00> : !cir.double
// CHECK-NEXT: %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !cir.double -> !cir.complex<!cir.double>
```

I'm assuming that the result of `cir.complex.create` is then stored to the global, right? If so, this seems good enough already. Am I missing something?

> However this may be a bit hard to implement during CIRGen since this assignment is basically an assignment between two complex values. The upstream CodeGen tracks the real and imaginary parts of a complex value individually so it can generate such code directly.

Can you elaborate on what upstream CodeGen does? I'm not sure I follow.

> In CIR we may use some canonicalization passes to do this job?

It's always a good option, not sure for this specific example there's much we could do though (I guess it depends on the reply above).

Lancern replied:
> I'm assuming that the result of cir.complex.create is then stored to the global, right? If so, this seems good enough already.

Yes. The result of `cir.complex.create` is then stored into the global.

> Can you elaborate on what upstream CodeGen does?

Let's say we're trying to CodeGen the following assignment statement:

```cpp
// int _Complex c;
c = 3i;
```

The statement contains two sub-expressions: the lhs is an lvalue that indicates the global variable `c`, and the rhs is a prvalue that represents a value of type `int _Complex`. To CodeGen the assignment statement, the CodeGen module will recursively CodeGen these two sub-expressions and then emit some store instructions to actually do the assignment.

When CodeGen-ing the rhs, the sub-expression CodeGen will yield a pair of LLVM values `std::pair<llvm::Value *, llvm::Value *>` to the statement CodeGen. This is how the upstream CodeGen represent a prvalue of complex type: the first element of the pair is the real part of the complex prvalue, and the second element is the imaginary part of the complex prvalue. When the statement CodeGen gets the pair of values, it then emits two store instructions to store the real and imaginary part into the global variable `c`.

In CIR, we take a different approach for CIRGen-ing prvalues of complex types. Instead of yielding a pair of values, CIRGen just yields a single `mlir::Value` that represents the whole complex value. Thus it would be more natural to successively generate a single store instruction to store the whole complex value into the global variable, rather than storing the real and imaginary parts separately.

Lancern replied:
A side note here: the literal `3i` is a complex value whose real part is implicitly 0. Thus one cannot just CodeGen / CIRGen the storing of the imaginary part; the storing of the implicit real part is also necessary.

bcardosolopes commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+//        CPP: cir.func @_Z9list_initv()
+//      CHECK:   %[[#REAL:]] = cir.const #cir.fp<1.000000e+00> : !cir.double
+// CHECK-NEXT:   %[[#IMAG:]] = cir.const #cir.fp<2.000000e+00> : !cir.double
+// CHECK-NEXT:   %{{.+}} = cir.complex.create %[[#REAL]], %[[#IMAG]] : !cir.double -> !cir.complex<!cir.double>
```
```
  %0 = cir.const #cir.fp<1.000000e+00> : !cir.double
  %1 = cir.const #cir.fp<2.000000e+00> : !cir.double
  %2 = cir.complex.create %0, %1 : !cir.complex<!cir.double>
  cir.store %2, ... : !cir.ptr<!cir.complex<!cir.double>>
```

I prefer keeping the `#cir.complex` to be used directly with `cir.const`, it composes well with the rest of CIR consts. By applying that change, `cir.complex.create` becomes only useful for "merging" two non-constants fp/ints in order to build a `!cir.complex`. But this brings the question if this is really useful, two possible approaches:

1. We remove it in favor of "get_real/get_imag + store to each", which reuses existing ops. A single store would still be used for when `cir.const` creates the complex type, which also brings some inconsistency.
2. We keep it. Useful to "join" two non-constants into a complex type so that we only have one final store instead of two, theoretically easier to track on analysis. The downside is having an extra op to build and later lower.

Perhaps we should do (2) because it's easier to translate to MLIR's complex dialect?

Lancern commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+#ifdef __cplusplus__
+#define VOID
+#else
+#define VOID void
```
Updated.

Lancern commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+//      CHECK: }
+
+void get_real(VOID) {
+  double *r1 = &__real__ c;
```
Updated.

Lancern commented on clang/test/CIR/CodeGen/complex.c:
```diff
@@ -0,0 +1,130 @@
+
+void list_init(VOID) {
+  double _Complex c1 = {1.0, 2.0};
+  int _Complex c2 = {1, 2};
```
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenFunction.h:
```diff
@@ -582,6 +582,22 @@ class CIRGenFunction : public CIRGenTypeCache {
 
   void buildAggExpr(const clang::Expr *E, AggValueSlot Slot);
 
+  /// buildComplexExpr - Emit the computation of the specified expression of
```
We should always remove repetition of method names in the comments, like "buildComplexExpr - " above.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprComplex.cpp:
```diff
@@ -0,0 +1,473 @@
+  //                               Utilities
+  //===--------------------------------------------------------------------===//
+
+  /// EmitLoadOfLValue - Given an expression with complex type that represents a
```
Same as last comment, please remove these extra method names everywhere you have them on comments.

Lancern replied:
Updated.

bcardosolopes reviewed (changes requested):
Thanks for revamping this, let's try to land it sooner than later. There are a few comments that seem like they didn't get address, can you update the ones that still need tackling versus the ones you consider done (I just solved a bunch but got unsure about few others).

> Hi Bruno, thanks for your reviews here. The problem with `#cir.complex` is that there is no literal expression in C that corresponds to it directly. One may think that the initializer list expression `{real, imag}` should be lowered to a `#cir.complex` but this could only happen when `real` and `imag` are both constant expressions. Otherwise you have no choice but fallback to `cir.complex.create`. Of course we could make `{real, imag}` lower to `cir.const #cir.complex` by trying to constant-evaluate the two elements during CIRGen.

Ok, I was thinking we could have a fold operation of sorts to canonicalize this, but it could come as a next step after we land this.

> `#cir.imag`, on the other hand, corresponds to the literal expression `1i`, `2i`, etc. In clang these literal expressions are of complex type and this is exactly what I'm trying to model in CIR here.

I understand that and I think it looks nice. However, I don't see any advantage of this representation brings - it's not even used in the tests (at least not tested AFAICT) - can you perhaps make the case? Perhaps it's something that can be later introduced if it proves itself worth for some more concrete reason?

Lancern commented:
Rebased onto the latest `main`.

Lancern commented:
I agree that `#cir.imag` looks inconsistent and not promising. Removed it from this PR, and now imaginary literals such as `1i` is lowered to three ops: two `cir.const` and one `cir.complex.create`:

```cpp
int _Complex c = 3i;
// %0 = cir.const #cir.int<0> : !s32i
// %1 = cir.const #cir.int<3> : !s32i
// %c = cir.complex.create %0, %1 : !s32i -> !cir.complex<!s32i>
```

Lancern commented:
> Ok, I was thinking we could have a fold operation of sorts to canonicalize this

This looks nice. I'll add this in another PR once this PR lands!

bcardosolopes commented:
> This looks nice. I'll add this in another PR once this PR lands!

Sounds good!

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1174,6 +1174,90 @@ def BinOpOverflowOp : CIR_Op<"binop.overflow", [Pure, SameTypeOperands]> {
+def ComplexImagPtrOp : CIR_Op<"complex.imag_ptr", [Pure]> {
+  let summary = "Extract the imaginary part of a complex value";
+  let description = [{
+    `cir.complex.get_real` operation takes a pointer operand that points to a
```
typo: `cir.complex.get_real` -> `complex.imag_ptr`

Lancern replied:
Updated.

bcardosolopes reviewed (approved):
Thanks for the hard work here, I'm glad to see this landing.

LGTM pending any new information from question in the comments!

Lancern commented:
Rebased onto the latest `main`.

Lancern commented:
Rebased onto the latest `main`.

bcardosolopes commented:
Pending replies to the questions asked.

Lancern commented:
> Pending replies to the questions asked.

It seems to me that all questions should be resolved? Am I missing some questions? GitHub does not show my response to your latest question, it's here: https://github.com/llvm/clangir/pull/513#discussion_r1649904257

bcardosolopes commented:
> It seems to me that all questions should be resolved? Am I missing some questions? GitHub does not show my response to your latest question in the thread, it's here: [#513 (comment)](https://github.com/llvm/clangir/pull/513#discussion_r1649904257)

Perfect, got confused by github! All seems good then, thanks for all the hard work here!
